src/core/common/tlvs.cpp

```diff
@@ -66,28 +66,33 @@ otError Tlv::GetOffset(const Message &aMessage, uint8_t aType, uint16_t &aOffset
     uint16_t end = aMessage.GetLength();
     Tlv tlv;
 
-    while (offset < end)
+    while (offset + sizeof(tlv) <= end)
     {
-        aMessage.Read(offset, sizeof(Tlv), &tlv);
+        uint16_t length;
 
-        // skip extended TLV
-        if (tlv.GetLength() == kExtendedLength)
-        {
-            uint16_t length = 0;
+        aMessage.Read(offset, sizeof(tlv), &tlv);
+        length = sizeof(tlv);
 
-            offset += sizeof(tlv);
-            aMessage.Read(offset, sizeof(length), &length);
-            offset += sizeof(length) + HostSwap16(length);
-        }
-        else if (tlv.GetType() == aType && (offset + sizeof(tlv) + tlv.GetLength()) <= end)
+        if (tlv.GetLength() != kExtendedLength)
         {
-            aOffset = offset;
-            ExitNow(error = OT_ERROR_NONE);
+            length += tlv.GetLength();
         }
         else
         {
-            offset += sizeof(tlv) + tlv.GetLength();
+            VerifyOrExit(offset + length + sizeof(length) <= end);
+            aMessage.Read(offset + length, sizeof(length), &length);
+            length += HostSwap16(length);
```

Comment:
I think we need to add `sizeof(tlv)` to `length` after reading it from `aMessage` here.... (we overwrite the original value of `length = sizeof(tlv);` when we read it from message)